cyipopt/utils.py

```diff
@@ -91,3 +91,33 @@ def generate_deprecation_warning_msg(what,
            f"deprecated in CyIpopt. Please replace all uses and use "
            f"'{new_name}' going forward.")
     return msg
+
+
+class FunctionWithApproxJacobian(object):
+    def __init__(self, func, epsilon):
+        self._func = func
+        self.epsilon = epsilon
+        self.value_cache = {}
+
+    def __call__(self, x, *args, **kwargs):
+        key = tuple(x)
+        if not key in self.value_cache:
+            value = self._func(x, *args, **kwargs)
+            if np.any(np.isnan(value)):
+                print("Warning! nan function value encountered at {0}".format(x))
```

Comment:
Ideally should raise a warning rather than just printing a warning-related message.